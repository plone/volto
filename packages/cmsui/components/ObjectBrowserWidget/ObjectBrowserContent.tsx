import React from 'react';
import { Button, Input } from '@plone/components/quanta';
import { SearchIcon, CloseIcon } from '@plone/components/Icons';
import { useTranslation } from 'react-i18next';

interface ObjectBrowserContentProps {
  searchMode: boolean;
  setSearchMode: (mode: boolean) => void;
  loading?: boolean;
  onSearchChange?: (text: string) => void;
  showSearch?: boolean;
  children: React.ReactNode;
}

export const ObjectBrowserContent = ({
  searchMode,
  setSearchMode,
  loading = false,
  onSearchChange,
  showSearch = true,
  children,
}: ObjectBrowserContentProps) => {
  const { t } = useTranslation();

  const handleSearchToggle = () => {
    setSearchMode(!searchMode);
    if (searchMode && onSearchChange) {
      onSearchChange(''); // Clear search when closing
    }
  };

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    if (onSearchChange) {
      onSearchChange(event.target.value);
    }
  };

  return (
    <div className="flex h-full flex-col">
      {showSearch && (
        <div className="border-b p-4">
          <div className="flex items-center gap-2">
            <Button
              variant="neutral"
              size="S"
              type="button"
              onPress={handleSearchToggle}
              aria-label={
                searchMode
                  ? t('cmsui.objectbrowserwidget.closeSearch')
                  : t('cmsui.objectbrowserwidget.openSearch')
              }
            >
              {searchMode ? <CloseIcon size="sm" /> : <SearchIcon size="sm" />}
            </Button>
            {searchMode && (
              <Input
                placeholder={t('cmsui.objectbrowserwidget.searchPlaceholder')}
                onChange={handleInputChange}
                autoFocus
                className="flex-1"
              />
            )}
          </div>
        </div>
      )}
      <div className="flex-1 overflow-hidden">
        {loading ? (
          <div className="flex h-full items-center justify-center">
            <div className="text-center text-gray-500">
              {t('cmsui.objectbrowserwidget.loading')}
            </div>
          </div>
        ) : (
          children
        )}
      </div>
    </div>
  );
};

ObjectBrowserContent.displayName = 'ObjectBrowserContent';
