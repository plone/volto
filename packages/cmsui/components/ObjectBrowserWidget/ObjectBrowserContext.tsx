'use client';

import { useFetcher } from 'react-router';
import {
  useCallback,
  useMemo,
  useState,
  useEffect,
  createContext,
  useContext,
  useId,
} from 'react';
import { useDebounceValue } from 'usehooks-ts';
import type { Selection, Key } from 'react-aria-components';
import type { Brain } from '@plone/types';
import {
  buildObjectBrowserUrl,
  processSelection,
  initializeSelectedKeys,
  type ObjectBrowserWidgetMode,
  useAccumulatedItems,
} from './utils';
import type { loader } from '../../routes/objectBrowserWidget';
import {
  ObjectBrowserNavigationProvider,
  useObjectBrowserNavigation,
} from './ObjectBrowserNavigationContext';

type PatternOptions = {
  maximumSelectionSize?: number;
  selectableTypes?: string[];
  basePath?: string;
  currentPath?: string;
} & Record<string, any>;

export interface UseObjectBrowserConfig {
  mode?: ObjectBrowserWidgetMode;
  widgetOptions?: {
    pattern_options?: PatternOptions;
  };
  selectedAttrs?: Array<keyof Brain>;
  // TODO: Also add blockchange/slate signature compat
  onChange?: (selected: Partial<Brain>[]) => void;
  defaultValue?: Brain[];
  title?: string;
  initialPath?: string;
}

// Hook interno che viene chiamato DOPO che il navigation context è disponibile
const useObjectBrowserInternal = (config: UseObjectBrowserConfig = {}) => {
  const {
    mode = 'multiple',
    widgetOptions = {},
    selectedAttrs = ['@id', 'title', 'description', '@type', 'UID'],
    onChange,
    defaultValue = [],
    title,
  } = config;
  const ariaControlsId = useId();
  const [open, setOpen] = useState(false);
  const [searchMode, setSearchMode] = useState(false);
  const fetcher = useFetcher<typeof loader>();
  const knownItems = useAccumulatedItems(fetcher, 'results');
  const [SearchableText, setSearchableText] = useDebounceValue<string>('', 350);
  const [selectedKeys, setSelectedKeys] = useState<
    Set<{ id: string; title: string }>
  >(() => initializeSelectedKeys(defaultValue));
  const { currentPath } = useObjectBrowserNavigation();

  // Funzioni esatte dal widget originale
  const filterBrainAttributes = useCallback(
    (brains: Brain[]) => {
      return brains.map(
        (item) =>
          Object.fromEntries(
            selectedAttrs
              .filter((attr) => attr in item)
              .map((attr) => [attr, item[attr]]),
          ) as Partial<Brain>,
      );
    },
    [selectedAttrs],
  );

  const handleSelectionChange = useCallback(
    (keys: Selection) => {
      const items = knownItems ?? [];
      const selected = processSelection(keys, items);

      const newSelectedKeys = new Set(
        selected.map((item) => ({ id: item['@id'], title: item.title })),
      );
      setSelectedKeys(newSelectedKeys);

      const filteredSelected = filterBrainAttributes(selected);
      onChange?.(filteredSelected);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [knownItems],
  );

  const handleRemove = useCallback(
    (keys: Set<Key>) => {
      const keysToRemove = new Set(keys);
      const newSelected = Array.from(selectedKeys).filter(
        (item) => !keysToRemove.has(item.id),
      );
      setSelectedKeys(new Set(newSelected));

      // Get complete Brain objects instead of just id/title
      const items = knownItems ?? [];
      const selectedBrains = newSelected
        .map(({ id }) => items.find((item) => item['@id'] === id))
        .filter((item): item is Brain => item !== undefined);

      const filteredSelected = filterBrainAttributes(selectedBrains);
      onChange?.(filteredSelected);
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [selectedKeys, knownItems],
  );

  const handleSearchInputChange = useCallback(
    (e: React.ChangeEvent<HTMLInputElement>) => {
      const value = e.target.value;
      setSearchableText(value);
    },
    [setSearchableText],
  );

  function reset() {
    setSearchableText('');
    fetcher.load('/reset-fetcher');
  }

  async function loadData(currentPath?: string, SearchableText?: string) {
    const url = buildObjectBrowserUrl(currentPath, SearchableText);
    if (!url) return;
    await fetcher.load(url);
  }

  useEffect(() => {
    if (searchMode && SearchableText.trim() === '') reset();
    else loadData(currentPath, SearchableText);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [currentPath, searchMode, SearchableText]);

  const items = useMemo(
    () => fetcher?.data?.results?.items ?? [],
    [fetcher.data],
  );

  const breadcrumbs = useMemo(
    () => fetcher?.data?.breadcrumbs?.items ?? [],
    [fetcher.data],
  );

  const loading = fetcher.state === 'loading';
  const selectedItems = Array.from(selectedKeys).map((item) => item.id);
  return {
    // State
    open,
    setOpen,
    searchMode,
    setSearchMode,
    selectedKeys,
    setSelectedKeys,
    SearchableText,
    setSearchableText,
    items,
    breadcrumbs,
    knownItems,
    loading,
    fetcher,
    selectedItems,

    ariaControlsId,
    // Actions
    handleSelectionChange,
    handleRemove,
    handleSearchInputChange,
    loadData,
    reset,
    filterBrainAttributes,

    // Config
    mode,
    selectedAttrs,
    title,
    widgetOptions,
  };
};

export type UseObjectBrowserReturn = ReturnType<
  typeof useObjectBrowserInternal
>;

const ObjectBrowserContext = createContext<UseObjectBrowserReturn | null>(null);
// Componente interno che usa l'hook dopo che il navigation context è disponibile
function ObjectBrowserInternalProvider({
  children,
  config,
}: {
  children: React.ReactNode;
  config: UseObjectBrowserConfig;
}) {
  const objectBrowser = useObjectBrowserInternal(config);

  const contextValue = useMemo(() => objectBrowser, [objectBrowser]);

  return (
    <ObjectBrowserContext.Provider value={contextValue}>
      {children}
    </ObjectBrowserContext.Provider>
  );
}

// Provider principale che gestisce l'ordine corretto dei provider annidati
export function ObjectBrowserProvider({
  children,
  config,
}: {
  children: React.ReactNode;
  config: UseObjectBrowserConfig;
}) {
  const { initialPath } = config;

  return (
    <ObjectBrowserNavigationProvider initialPath={initialPath}>
      <ObjectBrowserInternalProvider config={config}>
        {children}
      </ObjectBrowserInternalProvider>
    </ObjectBrowserNavigationProvider>
  );
}

export const useObjectBrowserContext = (): UseObjectBrowserReturn => {
  const context = useContext(ObjectBrowserContext);
  if (!context) {
    throw new Error(
      'useObjectBrowserContext must be used within an ObjectBrowserProvider',
    );
  }
  return context;
};

ObjectBrowserProvider.displayName = 'ObjectBrowserProvider';
