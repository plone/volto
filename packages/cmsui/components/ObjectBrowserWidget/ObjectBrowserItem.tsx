import React from 'react';
import { GridListItem, Button } from '@plone/components/quanta';
import { CheckboxIcon, ChevronrightIcon } from '@plone/components/Icons';
import { flattenToAppURL } from '@plone/helpers';
import { useTranslation } from 'react-i18next';
import { getItemLabel, getContentIcon } from './utils';
import type { Brain } from '@plone/types';
import { tv } from 'tailwind-variants';

export const itemVariants = tv({
  base: 'group',
  variants: {
    viewMode: {
      list: 'data-[selectable=false]:bg-quanta-silver data-[selectable=false]:text-iron border-l-4 data-[selectable=false]:cursor-not-allowed',
      grid: 'pb-2 hover:cursor-pointer hover:bg-transparent',
    },
    workflow: {
      private: '',
      published: '',
      pending: '',
    },
  },
  compoundVariants: [
    // List view with workflow border colors
    {
      viewMode: 'list',
      workflow: 'private',
      class: 'border-l-red-500',
    },
    {
      viewMode: 'list',
      workflow: 'published',
      class: 'border-l-blue-500',
    },
    {
      viewMode: 'list',
      workflow: 'pending',
      class: 'border-l-yellow-500',
    },
  ],
});

export interface ObjectBrowserItemProps {
  item: Brain;
  onNavigate: (item: Brain) => void;
  viewMode: boolean;
}

export function ObjectBrowserItem({
  item,
  viewMode,
  onNavigate,
}: ObjectBrowserItemProps) {
  const { t } = useTranslation();
  const ContentIcon = getContentIcon(
    item['@type'] || item.portal_type || 'Document',
    item.is_folderish || false,
  );

  return viewMode ? (
    // Grid view - no workflow border, content icon
    <div className="relative flex w-36 flex-1 flex-col items-center gap-2 p-0">
      <div
        className={`
          absolute top-2 left-2 hidden rounded-full border border-solid border-white bg-white
          leading-0 text-quanta-sapphire
          group-selected:block
        `}
      >
        <CheckboxIcon className="h-6 w-6" />
      </div>
      <div
        className={`
          flex min-h-36 min-w-36 flex-col items-center justify-center rounded-md bg-quanta-snow
          text-center
        `}
      >
        {item.image_field && item.image_scales ? (
          <img
            className="h-full w-full rounded-md object-contain"
            src={flattenToAppURL(
              `${item['@id']}/${item.image_scales.image[0].scales.preview?.download}`,
            )}
            alt={item.title || ''}
          />
        ) : (
          <div className="flex h-16 w-16 items-center justify-center text-quanta-pigeon">
            <ContentIcon size="xl" />
          </div>
        )}
      </div>
      <span className="text-center text-sm leading-tight font-medium text-quanta-cobalt">
        {item.title}
      </span>
    </div>
  ) : (
    // List view - with workflow border color and icons
    <div
      className={`
        flex h-full w-full items-center text-quanta-cobalt
        group-data-[selectable=false]:text-quanta-iron
      `}
    >
      <div className="flex min-w-0 flex-1 items-center gap-3">
        <ContentIcon size="sm" className="text-quanta-pigeon" />
        <span className="truncate font-medium">{item.title}</span>
      </div>

      {/* Fixed width column for consistent arrow alignment */}
      <div className="flex w-8 flex-shrink-0 justify-center">
        {item.is_folderish && (
          <Button
            variant="icon"
            type="button"
            aria-label={t('cmsui.objectbrowserwidget.itemNavigateTo', {
              title: item.title,
            })}
            className="rounded-none border-none bg-transparent"
            onPress={() => onNavigate(item)}
          >
            <ChevronrightIcon size="sm" />
          </Button>
        )}
      </div>
    </div>
  );
}
