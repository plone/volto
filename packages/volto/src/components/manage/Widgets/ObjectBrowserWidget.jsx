/**
 * ObjectBrowserWidget component.
 * @module components/manage/Widgets/ObjectBrowserWidget
 */

import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { compose } from 'redux';
import { compact, isArray, isEmpty, remove } from 'lodash';
import { connect } from 'react-redux';
import { Label, Popup, Button } from 'semantic-ui-react';
import {
  flattenToAppURL,
  isInternalURL,
  isUrl,
  normalizeUrl,
  removeProtocol,
} from '@plone/volto/helpers/Url/Url';
import { searchContent } from '@plone/volto/actions/search/search';
import withObjectBrowser from '@plone/volto/components/manage/Sidebar/ObjectBrowser';
import { defineMessages, injectIntl } from 'react-intl';
import Icon from '@plone/volto/components/theme/Icon/Icon';
import FormFieldWrapper from '@plone/volto/components/manage/Widgets/FormFieldWrapper';

import navTreeSVG from '@plone/volto/icons/nav.svg';
import clearSVG from '@plone/volto/icons/clear.svg';
import homeSVG from '@plone/volto/icons/home.svg';
import aheadSVG from '@plone/volto/icons/ahead.svg';
import blankSVG from '@plone/volto/icons/blank.svg';
import { withRouter } from 'react-router';

const messages = defineMessages({
  placeholder: {
    id: 'No items selected',
    defaultMessage: 'No items selected',
  },
  edit: {
    id: 'Edit',
    defaultMessage: 'Edit',
  },
  delete: {
    id: 'Delete',
    defaultMessage: 'Delete',
  },
  openObjectBrowser: {
    id: 'Open object browser',
    defaultMessage: 'Open object browser',
  },
});

/**
 * ObjectBrowserWidget component class.
 * @class ObjectBrowserWidget
 * @extends Component
 */
export class ObjectBrowserWidgetComponent extends Component {
  /**
   * Property types.
   * @property {Object} propTypes Property types.
   * @static
   */
  static propTypes = {
    id: PropTypes.string.isRequired,
    title: PropTypes.string.isRequired,
    description: PropTypes.string,
    mode: PropTypes.string, // link, image, multiple
    return: PropTypes.string, // single, multiple
    initialPath: PropTypes.string,
    required: PropTypes.bool,
    error: PropTypes.arrayOf(PropTypes.string),
    value: PropTypes.oneOfType([
      PropTypes.arrayOf(PropTypes.object),
      PropTypes.object,
    ]),
    onChange: PropTypes.func.isRequired,
    openObjectBrowser: PropTypes.func.isRequired,
    allowExternals: PropTypes.bool,
    placeholder: PropTypes.string,
  };

  /**
   * Default properties
   * @property {Object} defaultProps Default properties.
   * @static
   */
  static defaultProps = {
    description: null,
    required: false,
    error: [],
    value: [],
    mode: 'multiple',
    return: 'multiple',
    initialPath: '',
    allowExternals: false,
  };

  state = {
    manualLinkInput: '',
    validURL: false,
  };

  constructor(props) {
    super(props);
    this.selectedItemsRef = React.createRef();
    this.placeholderRef = React.createRef();
  }
  renderLabel(item) {
    const href = item['@id'];
    return (
      <Popup
        key={flattenToAppURL(href)}
        content={
          <div style={{ display: 'flex' }}>
            {isInternalURL(href) ? (
              <Icon name={homeSVG} size="18px" />
            ) : (
              <Icon name={blankSVG} size="18px" />
            )}
            &nbsp;
            {flattenToAppURL(href)}
          </div>
        }
        trigger={
          <Label>
            <div className="item-title">{item.title}</div>
            <div>
              {this.props.mode === 'multiple' && (
                <Icon
                  name={clearSVG}
                  size="12px"
                  className="right"
                  onClick={(event) => {
                    event.preventDefault();
                    this.removeItem(item);
                  }}
                />
              )}
            </div>
          </Label>
        }
      />
    );
  }

  removeItem = (item) => {
    let value = [...this.props.value];
    remove(value, function (_item) {
      return _item['@id'] === item['@id'];
    });
    this.props.onChange(this.props.id, value);
  };

  onChange = (item) => {
    let value =
      this.props.mode === 'multiple' && this.props.value
        ? [...this.props.value]
        : [];
    value = value.filter((item) => item != null);
    const maxSize =
      this.props.widgetOptions?.pattern_options?.maximumSelectionSize || -1;
    if (maxSize === 1 && value.length === 1) {
      value = []; //enable replace of selected item with another value, if maxsize is 1
    }
    let exists = false;
    let index = -1;
    value.forEach((_item, _index) => {
      if (flattenToAppURL(_item['@id']) === flattenToAppURL(item['@id'])) {
        exists = true;
        index = _index;
      }
    });
    //find(value, {
    //   '@id': flattenToAppURL(item['@id']),
    // });
    if (!exists) {
      // add item
      // Check if we want to filter the attributes of the selected item
      let resultantItem = item;
      if (this.props.selectedItemAttrs) {
        const allowedItemKeys = [
          ...this.props.selectedItemAttrs,
          // Add the required attributes for the widget to work
          '@id',
          'title',
        ];
        resultantItem = Object.keys(item)
          .filter((key) => allowedItemKeys.includes(key))
          .reduce((obj, key) => {
            obj[key] = item[key];
            return obj;
          }, {});
      }
      // Add required @id field, just in case
      resultantItem = { ...resultantItem, '@id': item['@id'] };
      value.push(resultantItem);
      if (this.props.return === 'single') {
        this.props.onChange(this.props.id, value[0]);
      } else {
        this.props.onChange(this.props.id, value);
      }
    } else {
      //remove item
      value.splice(index, 1);
      this.props.onChange(this.props.id, value);
    }
  };

  onManualLinkInput = (e) => {
    this.setState({ manualLinkInput: e.target.value });
    if (this.validateManualLink(e.target.value)) {
      this.setState({ validURL: true });
    } else {
      this.setState({ validURL: false });
    }
  };

  validateManualLink = (url) => {
    if (this.props.allowExternals) {
      return isUrl(url);
    } else {
      return isInternalURL(url);
    }
  };

  onSubmitManualLink = () => {
    if (this.validateManualLink(this.state.manualLinkInput)) {
      if (isInternalURL(this.state.manualLinkInput)) {
        const link = this.state.manualLinkInput;
        // convert it into an internal on if possible
        this.props
          .searchContent(
            '/',
            {
              'path.query': flattenToAppURL(this.state.manualLinkInput),
              'path.depth': '0',
              sort_on: 'getObjPositionInParent',
              metadata_fields: '_all',
              b_size: 1000,
            },
            `${this.props.block}-${this.props.mode}`,
          )
          .then((resp) => {
            if (resp.items?.length > 0) {
              this.onChange(resp.items[0]);
            } else {
              this.props.onChange(this.props.id, [
                {
                  '@id': flattenToAppURL(link),
                  title: removeProtocol(link),
                },
              ]);
            }
          });
      } else {
        this.props.onChange(this.props.id, [
          {
            '@id': normalizeUrl(this.state.manualLinkInput),
            title: removeProtocol(this.state.manualLinkInput),
          },
        ]);
      }
      this.setState({ validURL: true, manualLinkInput: '' });
    }
  };

  onKeyDownManualLink = (e) => {
    if (e.key === 'Enter') {
      e.preventDefault();
      e.stopPropagation();
      this.onSubmitManualLink();
    } else if (e.key === 'Escape') {
      e.preventDefault();
      e.stopPropagation();
      // TODO: Do something on ESC key
    }
  };

  showObjectBrowser = (ev) => {
    ev.preventDefault();
    this.props.openObjectBrowser({
      mode: this.props.mode,
      currentPath: this.props.initialPath || this.props.location.pathname,
      propDataName: 'value',
      onSelectItem: (url, item) => {
        this.onChange(item);
      },
      selectableTypes:
        this.props.widgetOptions?.pattern_options?.selectableTypes ||
        this.props.selectableTypes,
      maximumSelectionSize:
        this.props.widgetOptions?.pattern_options?.maximumSelectionSize ||
        this.props.maximumSelectionSize,
    });
  };

  handleSelectedItemsRefClick = (e) => {
    if (this.props.isDisabled) {
      return;
    }

    if (
      e.target.contains(this.selectedItemsRef.current) ||
      e.target.contains(this.placeholderRef.current)
    ) {
      this.showObjectBrowser(e);
    }
  };

  /**
   * Render method.
   * @method render
   * @returns {string} Markup for the component.
   */
  render() {
    const { id, description, fieldSet, value, mode, onChange, isDisabled } =
      this.props;

    let items = compact(!isArray(value) && value ? [value] : value || []);

    let icon =
      mode === 'multiple' || items.length === 0 ? navTreeSVG : clearSVG;
    let iconAction =
      mode === 'multiple' || items.length === 0
        ? this.showObjectBrowser
        : (e) => {
            e.preventDefault();
            onChange(id, this.props.return === 'single' ? null : []);
          };

    return (
      <FormFieldWrapper
        {...this.props}
        className={description ? 'help text' : 'text'}
      >
        <div
          className="objectbrowser-field"
          aria-labelledby={`fieldset-${
            fieldSet || 'default'
          }-field-label-${id}`}
        >
          <div
            className="selected-values"
            onClick={this.handleSelectedItemsRefClick}
            onKeyDown={this.handleSelectedItemsRefClick}
            role="searchbox"
            tabIndex={0}
            ref={this.selectedItemsRef}
          >
            {items.map((item) => this.renderLabel(item))}

            {items.length === 0 && this.props.mode === 'multiple' && (
              <div className="placeholder" ref={this.placeholderRef}>
                {this.props.placeholder ??
                  this.props.intl.formatMessage(messages.placeholder)}
              </div>
            )}
            {this.props.allowExternals &&
              items.length === 0 &&
              this.props.mode !== 'multiple' && (
                <input
                  onKeyDown={this.onKeyDownManualLink}
                  onChange={this.onManualLinkInput}
                  value={this.state.manualLinkInput}
                  placeholder={
                    this.props.placeholder ??
                    this.props.intl.formatMessage(messages.placeholder)
                  }
                />
              )}
          </div>
          {this.state.manualLinkInput && isEmpty(items) && (
            <Button.Group>
              <Button
                basic
                className="cancel"
                onClick={(e) => {
                  e.stopPropagation();
                  this.setState({ manualLinkInput: '' });
                }}
              >
                <Icon name={clearSVG} size="18px" color="#e40166" />
              </Button>
              <Button
                basic
                primary
                disabled={!this.state.validURL}
                onClick={(e) => {
                  e.stopPropagation();
                  this.onSubmitManualLink();
                }}
              >
                <Icon name={aheadSVG} size="18px" />
              </Button>
            </Button.Group>
          )}
          {!this.state.manualLinkInput && (
            <Button
              aria-label={this.props.intl.formatMessage(
                messages.openObjectBrowser,
              )}
              onClick={iconAction}
              className="action"
              disabled={isDisabled}
            >
              <Icon name={icon} size="18px" />
            </Button>
          )}
        </div>
      </FormFieldWrapper>
    );
  }
}

const ObjectBrowserWidgetMode = (mode) =>
  compose(
    injectIntl,
    withObjectBrowser,
    withRouter,
    connect(null, { searchContent }),
  )((props) => <ObjectBrowserWidgetComponent {...props} mode={mode} />);
export { ObjectBrowserWidgetMode };
export default compose(
  injectIntl,
  withObjectBrowser,
  withRouter,
  connect(null, { searchContent }),
)(ObjectBrowserWidgetComponent);
