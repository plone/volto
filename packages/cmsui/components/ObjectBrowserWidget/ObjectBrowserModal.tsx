'use client';

import { Dialog, Heading } from 'react-aria-components';
import { Modal } from '@plone/components';
import { Button, Input } from '@plone/components/quanta';
import { SearchIcon, CloseIcon } from '@plone/components/Icons';
import { ObjectBrowserWidgetBody } from './ObjectBrowserWidgetBody';
import { useObjectBrowserContext } from './ObjectBrowserContext';
import { useTranslation } from 'react-i18next';

export const ObjectBrowserModal = () => {
  const { t } = useTranslation();
  const {
    open,
    setOpen,
    searchMode,
    setSearchMode,
    setSearchableText,
    handleSearchInputChange,
    ariaControlsId,
    title,
  } = useObjectBrowserContext();
  return (
    <Modal
      isDismissable
      className={`
        data-[entering]:animate-slide-in
        data-[exiting]:animate-slide-out
        fixed top-0 right-0 bottom-0 w-[360px] border-l border-quanta-azure bg-quanta-air px-6 py-8
        text-black shadow-[rgba(0,0,0,0.1)_-8px_0px_20px] outline-none
      `}
      isOpen={open}
      onOpenChange={(isOpen) => setOpen(isOpen)}
    >
      <Dialog className="flex h-full flex-col overflow-hidden p-1">
        {!searchMode ? (
          <div className="flex items-center justify-between">
            <Heading slot="title" className="!mb-0 !text-2xl">
              {title || t('cmsui.objectbrowserwidget.dialogTitle')}
            </Heading>
            <div className="flex items-center gap-0.5">
              <Button
                variant="icon"
                onPress={() => {
                  setSearchMode(true);
                  setSearchableText('');
                }}
                type="button"
                aria-label={t('cmsui.objectbrowserwidget.openSearch')}
              >
                <SearchIcon />
              </Button>

              <Button
                slot="close"
                variant="icon"
                type="button"
                aria-label={t('cmsui.objectbrowserwidget.closeDialog')}
                onPress={() => setOpen(false)}
              >
                <CloseIcon />
              </Button>
            </div>
          </div>
        ) : (
          <div className="flex items-center gap-2">
            <Input
              onChange={handleSearchInputChange}
              className={'border-quanta rounded-md'}
              aria-controls={ariaControlsId}
              placeholder={t('cmsui.objectbrowserwidget.searchPlaceholder')}
            />
            <Button
              variant="icon"
              type="button"
              aria-label={t('cmsui.objectbrowserwidget.closeSearch')}
              onPress={() => {
                setSearchMode(false);
                setSearchableText('');
              }}
            >
              <CloseIcon />
            </Button>
          </div>
        )}
        <ObjectBrowserWidgetBody />
      </Dialog>
    </Modal>
  );
};

ObjectBrowserModal.displayName = 'ObjectBrowserModal';
