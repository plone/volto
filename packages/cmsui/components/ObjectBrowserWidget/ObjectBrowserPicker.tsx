// Componente per il caso "solo bottone + modal" senza tags
import React from 'react';
import { DialogTrigger } from 'react-aria-components';
import {
  useObjectBrowser,
  type UseObjectBrowserConfig,
} from './ObjectBrowserContext';
import { ObjectBrowserTrigger } from './ObjectBrowserTrigger';
import { ObjectBrowserModal } from './ObjectBrowserModal';

interface ObjectBrowserPickerProps extends UseObjectBrowserConfig {
  className?: string;
}

export const ObjectBrowserPicker = ({
  className,
  ...config
}: ObjectBrowserPickerProps) => {
  const {
    // State dal hook
    open,
    setOpen,
  } = useObjectBrowser(config);

  return (
    <div className={className}>
      <DialogTrigger isOpen={open} onOpenChange={setOpen}>
        <ObjectBrowserTrigger />
        <ObjectBrowserModal />
      </DialogTrigger>
    </div>
  );
};
