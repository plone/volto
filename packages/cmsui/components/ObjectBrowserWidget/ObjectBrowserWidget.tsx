import type { BaseFormFieldProps } from '../TextField/TextField';
import {
  Description,
  fieldBorderStyles,
  FieldError,
  Label,
} from '../Field/Field';
import { tv } from 'tailwind-variants';
import { focusRing } from '../utils';
import { useLoaderData } from 'react-router';
import type { loader as editLoader } from '../../routes/edit';
import { ObjectBrowserProvider } from './ObjectBrowserContext';
import { ObjectBrowserTags } from './ObjectBrowserTags';
import { ObjectBrowserTrigger } from './ObjectBrowserTrigger';
import { ObjectBrowserModal } from './ObjectBrowserModal';
import { useFocusRing, useId } from 'react-aria';

// TODO: better styling
const widgetStyles = tv({
  extend: focusRing,
  base: 'mx-1 flex items-center justify-between gap-2 rounded-md',
  variants: {
    isFocused: fieldBorderStyles.variants.isFocusWithin,
    isInvalid: fieldBorderStyles.variants.isInvalid,
    isDisabled: fieldBorderStyles.variants.isDisabled,
  },
});

interface ObjectBrowserWidgetProps extends BaseFormFieldProps {}
// TODO: interaction with plate and blocks schema
export function ObjectBrowserWidgetComponent(props: ObjectBrowserWidgetProps) {
  const { label, description, errorMessage } = props;
  const { isFocusVisible, focusProps } = useFocusRing();
  const id = useId();
  return (
    <div className="group mb-4 flex flex-col gap-1">
      {label && (
        <Label
          id={id}
          className={`
            not-group-data-invalid:not-group-data-readonly:has-[+div:focus]:text-quanta-sapphire
          `}
        >
          {label}
        </Label>
      )}
      {/* // TODO: maybe find a better way to use focus and leverage group focus styles */}
      <div
        {...focusProps}
        aria-labelledby={id}
        className={widgetStyles({
          isFocusVisible,
          isInvalid: !!props.errorMessage,
        })}
        role="group"
      >
        <ObjectBrowserTags />
        <ObjectBrowserTrigger>
          <ObjectBrowserModal />
        </ObjectBrowserTrigger>
      </div>
      {description && <Description>{description}</Description>}
      <FieldError>{errorMessage}</FieldError>
    </div>
  );
}

export function ObjectBrowserWidget(props: ObjectBrowserWidgetProps) {
  const { content } = useLoaderData<typeof editLoader>();
  const { label, description, errorMessage, ...rest } = props;
  return (
    <ObjectBrowserProvider config={{ ...rest, initialPath: content?.['@id'] }}>
      <ObjectBrowserWidgetComponent {...{ label, description, errorMessage }} />
    </ObjectBrowserProvider>
  );
}

ObjectBrowserWidget.displayName = 'ObjectBrowserWidget';
