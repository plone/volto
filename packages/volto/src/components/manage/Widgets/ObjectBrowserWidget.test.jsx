import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import ObjectBrowserWidget from './ObjectBrowserWidget';
import { MemoryRouter } from 'react-router-dom';

const mockStore = configureStore();
test('renders a objectBrowser widget component', () => {
  const store = mockStore({
    search: {},
    intl: {
      locale: 'en',
      messages: {},
    },
  });
  const component = renderer.create(
    <Provider store={store}>
      <MemoryRouter>
        <ObjectBrowserWidget
          id="my-field"
          title="My field"
          fieldSet="default"
          onChange={() => {}}
        />
      </MemoryRouter>
    </Provider>,
  );
  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});
