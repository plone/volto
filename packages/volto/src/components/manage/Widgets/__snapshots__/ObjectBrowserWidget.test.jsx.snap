// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders a objectBrowser widget component 1`] = `
<div
  className="inline field text field-wrapper-my-field"
>
  <div
    className="ui grid"
  >
    <div
      className="stretched row"
    >
      <div
        className="four wide column"
      >
        <div
          className="wrapper"
        >
          <label
            htmlFor="field-my-field"
            id="fieldset-default-field-label-my-field"
          >
            My field
          </label>
        </div>
      </div>
      <div
        className="eight wide column"
      >
        <div
          aria-labelledby="fieldset-default-field-label-my-field"
          className="objectbrowser-field"
        >
          <div
            className="selected-values"
            onClick={[Function]}
            onKeyDown={[Function]}
            role="searchbox"
            tabIndex={0}
          >
            <div
              className="placeholder"
            >
              No items selected
            </div>
          </div>
          
          <button
            aria-label="Open object browser"
            className="ui button action"
            onClick={[Function]}
          >
            <svg
              className="icon"
              dangerouslySetInnerHTML={
                Object {
                  "__html": undefined,
                }
              }
              onClick={null}
              style={
                Object {
                  "fill": "currentColor",
                  "height": "18px",
                  "width": "auto",
                }
              }
              viewBox=""
              xmlns=""
            />
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
